src/basic/elf-metadata.c

```diff
@@ -0,0 +1,123 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <dlfcn.h>
+#include <link.h>
+#include <string.h>
+
+/* This macro will turn off invocation of elf_build_id_get() by the log.c logging macros, so that we don't
+ * end up in a loop where we want to log, and the log function macros want to call us. */
+#define LOG_BUILD_ID_SUPPRESS
+
+#include "elf-metadata.h"
+#include "log.h"
+#include "macro.h"
+
+static thread_local struct {
+        const char *func;
+        elf_build_id build_id;
+} cache = {};
+
+const elf_build_id* elf_build_id_get(const char *func) {
+        struct link_map *m = NULL;
+        Dl_info info = {};
+        ElfW(Ehdr) *h;
+
+        /* This returns a static const structure referencing the build ID of the calling function. The return
+         * structure and the data its fields point to are valid as long as the relevant ELF module remains
+         * loaded. */
+
+        if (func && cache.func == func) {
+                /* We do some really basic caching: the caller is supposed to pass __func__ as argument, and
+                 * as long as that pointer doesn't change we'll return the same cached build ID. Note that we
+                 * compare by pointer here (!) since __func__ is defined to resolve to a static const char
+                 * array, and hence it's address should suffice for identifying whether we are still in the
+                 * same function. */
+
+                if (cache.build_id.size == 0) /* couldn't resolve (or in other words: we do negative caching too) */
+                        return NULL;
+
+                return &cache.build_id;
+        }
+
+        /* Look at the caller's address, and find the ELF module for it */
+        if (dladdr1(__builtin_return_address(0), &info, (void**) &m, RTLD_DL_LINKMAP) == 0) {
+                log_debug("Failed to resolve calling address with dladdr1(): %m");
+                goto fail;
+        }
+
+        /* This should be an ELF object */
+        h = (ElfW(Ehdr)*) info.dli_fbase;
+        if (!h ||
+            h->e_ident[0] != 0x7f ||
+            h->e_ident[1] != 'E' ||
+            h->e_ident[2] != 'L' ||
+            h->e_ident[3] != 'F') {
+                log_debug("Discovered object does not point to ELF header, ignoring.");
+                goto fail;
+        }
+
+        /* Let's iterate through all sections of this ELF object, and look for the PT_NOTE ones */
```

Comment:
libdwfelf.h from elfutils has `dwelf_elf_gnu_build_id()` which will give you the build-id given an `Elf *` reference - can't we use that instead of reimplementing the parsing?

Reply:
we want to know the build ID of our own ELF object we run in, and without dwarf or anything (because that is likely stripped).

libdwfeld is great, if you analyze foreign binaries or coredump, but when we want to do this inside of a binary then I think we better use the stuff glibc exports for this. Moreover the idea is to do this from *all* programs really, and I doubt we want to add that libdwfelf dep to every single binary....